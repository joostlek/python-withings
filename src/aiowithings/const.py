"""Asynchronous Python client for Withings."""
from __future__ import annotations

import logging

LOGGER = logging.getLogger(__package__)

STATUS_SUCCESS: set[int] = {0}

STATUS_AUTH_FAILED: set[int] = {100, 101, 102, 200, 401}

STATUS_INVALID_PARAMS: set[int] = {
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    211,
    212,
    213,
    216,
    217,
    218,
    220,
    221,
    223,
    225,
    227,
    228,
    229,
    230,
    234,
    235,
    236,
    238,
    240,
    241,
    242,
    243,
    244,
    245,
    246,
    247,
    248,
    249,
    250,
    251,
    252,
    254,
    260,
    261,
    262,
    263,
    264,
    265,
    266,
    267,
    271,
    272,
    275,
    276,
    283,
    284,
    285,
    286,
    287,
    288,
    290,
    293,
    294,
    295,
    297,
    300,
    301,
    302,
    303,
    304,
    321,
    323,
    324,
    325,
    326,
    327,
    328,
    329,
    330,
    331,
    332,
    333,
    334,
    335,
    336,
    337,
    338,
    339,
    340,
    341,
    342,
    343,
    344,
    345,
    346,
    347,
    348,
    349,
    350,
    351,
    352,
    353,
    380,
    381,
    382,
    400,
    501,
    502,
    503,
    504,
    505,
    506,
    509,
    510,
    511,
    523,
    532,
    3017,
    3018,
    3019,
}

STATUS_UNAUTHORIZED: set[int] = {214, 277, 2553, 2554, 2555}

STATUS_ERROR_OCCURRED: set[int] = {
    215,
    219,
    222,
    224,
    226,
    231,
    233,
    237,
    253,
    255,
    256,
    257,
    258,
    259,
    268,
    269,
    270,
    273,
    274,
    278,
    279,
    280,
    281,
    282,
    289,
    291,
    292,
    296,
    298,
    305,
    306,
    308,
    309,
    310,
    311,
    312,
    313,
    314,
    315,
    316,
    317,
    318,
    319,
    320,
    322,
    370,
    371,
    372,
    373,
    374,
    375,
    383,
    391,
    402,
    516,
    517,
    518,
    519,
    520,
    521,
    525,
    526,
    527,
    528,
    529,
    530,
    531,
    533,
    602,
    700,
    1051,
    1052,
    1053,
    1054,
    2551,
    2552,
    2556,
    2557,
    2558,
    2559,
    3000,
    3001,
    3002,
    3003,
    3004,
    3005,
    3006,
    3007,
    3008,
    3009,
    3010,
    3011,
    3012,
    3013,
    3014,
    3015,
    3016,
    3020,
    3021,
    3022,
    3023,
    3024,
    5000,
    5001,
    5005,
    5006,
    6000,
    6010,
    6011,
    9000,
    10000,
}

STATUS_TIMEOUT: set[int] = {522}
STATUS_BAD_STATE: set[int] = {524}
STATUS_TOO_MANY_REQUESTS: set[int] = {601}
