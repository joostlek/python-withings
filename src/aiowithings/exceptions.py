"""Asynchronous Python client for Withings."""


class WithingsError(Exception):
    """Generic exception."""


class WithingsConnectionError(WithingsError):
    """Withings connection exception."""


class WithingsAuthenticationFailedError(WithingsError):
    """Withings authentication failed exception."""


class WithingsInvalidParamsError(WithingsError):
    """Withings invalid params exception."""


class WithingsUnauthorizedError(WithingsError):
    """Withings unauthorized exception."""


class WithingsErrorOccurredError(WithingsError):
    """Withings error occurred exception."""


class WithingsBadStateError(WithingsError):
    """Withings bad state exception."""


class WithingsTooManyRequestsError(WithingsError):
    """Withings too many requests exception."""


class WithingsUnknownStatusError(WithingsError):
    """Withings unknown status exception."""
