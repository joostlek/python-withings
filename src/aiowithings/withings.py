"""Asynchronous Python client for Withings."""
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from importlib import metadata
from typing import TYPE_CHECKING, Any, Awaitable, Callable, cast

from aiohttp import ClientSession
from aiohttp.hdrs import METH_POST
import async_timeout
from yarl import URL

from .const import (
    STATUS_AUTH_FAILED,
    STATUS_BAD_STATE,
    STATUS_ERROR_OCCURRED,
    STATUS_INVALID_PARAMS,
    STATUS_SUCCESS,
    STATUS_TIMEOUT,
    STATUS_TOO_MANY_REQUESTS,
    STATUS_UNAUTHORIZED,
)
from .exceptions import (
    WithingsAuthenticationFailedError,
    WithingsBadStateError,
    WithingsConnectionError,
    WithingsError,
    WithingsErrorOccurredError,
    WithingsInvalidParamsError,
    WithingsTooManyRequestsError,
    WithingsUnauthorizedError,
    WithingsUnknownStatusError,
)
from .models import (
    Device,
    Goals,
    MeasurementGroup,
    MeasurementType,
    NotificationCategory,
    NotificationConfiguration,
    SleepDataFields,
    SleepSeries,
    SleepSummary,
    SleepSummaryDataFields,
)

if TYPE_CHECKING:
    from datetime import date, datetime

    from typing_extensions import Self


@dataclass
class WithingsClient:
    """Main class for handling connections with Withings."""

    session: ClientSession | None = None
    request_timeout: int = 10
    api_host: str = "wbsapi.withings.net"
    _token: str | None = None
    _close_session: bool = False
    refresh_token_function: Callable[[], Awaitable[str]] | None = None

    async def refresh_token(self) -> None:
        """Refresh token with provided function."""
        if self.refresh_token_function:
            self._token = await self.refresh_token_function()

    def authenticate(self, token: str) -> None:
        """Authenticate the user with a token."""
        self._token = token

    async def _request(
        self,
        uri: str,
        *,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Handle a request to Withings."""
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host=self.api_host,
            port=443,
        ).joinpath(uri)

        await self.refresh_token()

        headers = {
            "User-Agent": f"AioWithings/{version}",
            "Accept": "application/json, text/plain, */*",
            "Authorization": f"Bearer {self._token}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    METH_POST,
                    url,
                    headers=headers,
                    data=data,
                )
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to Withings"
            raise WithingsConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")

        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected response from Withings"
            raise WithingsError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        response_data = cast(dict[str, Any], await response.json())
        response_status = response_data.get("status", -1)
        if response_status in STATUS_SUCCESS:
            return cast(dict[str, Any], response_data.get("body"))
        error = response_data.get("error")
        if response_status in STATUS_AUTH_FAILED:
            raise WithingsAuthenticationFailedError(error)
        if response_status in STATUS_INVALID_PARAMS:
            raise WithingsInvalidParamsError(error)
        if response_status in STATUS_UNAUTHORIZED:
            raise WithingsUnauthorizedError(error)
        if response_status in STATUS_ERROR_OCCURRED:
            raise WithingsErrorOccurredError(error)
        if response_status in STATUS_TIMEOUT:
            raise WithingsConnectionError(error)
        if response_status in STATUS_BAD_STATE:
            raise WithingsBadStateError(error)
        if response_status in STATUS_TOO_MANY_REQUESTS:
            raise WithingsTooManyRequestsError(error)
        raise WithingsUnknownStatusError(error)

    async def get_devices(self) -> list[Device]:
        """Get devices."""
        response = await self._request("v2/user", data={"action": "getdevice"})
        return [Device.from_api(device) for device in response["devices"]]

    async def get_goals(self) -> Goals:
        """Get goals."""
        response = await self._request("v2/user", data={"action": "getgoals"})
        return Goals.from_api(response["goals"])

    async def _get_measurements(
        self,
        measurement_types: list[MeasurementType] | None,
        base_data: dict[str, Any],
    ) -> list[MeasurementGroup]:
        data = {**base_data, "action": "getmeas"}
        if measurement_types is not None:
            data["meastypes"] = ",".join(
                [str(measurement_type) for measurement_type in measurement_types],
            )
        response = await self._request("measure", data=data)
        return [
            MeasurementGroup.from_api(measurement_group)
            for measurement_group in response["measuregrps"]
        ]

    async def get_measurement_since(
        self,
        measurement_since: datetime,
        measurement_types: list[MeasurementType] | None = None,
    ) -> list[MeasurementGroup]:
        """Get all measurements since measurement_since."""
        return await self._get_measurements(
            measurement_types,
            {"lastupdate": measurement_since.timestamp()},
        )

    async def get_measurement_in_period(
        self,
        start_date: datetime,
        end_date: datetime,
        measurement_types: list[MeasurementType] | None = None,
    ) -> list[MeasurementGroup]:
        """Get all measurements measured since start date and until end date."""
        return await self._get_measurements(
            measurement_types,
            {"startdate": start_date.timestamp(), "enddate": end_date.timestamp()},
        )

    async def get_sleep(
        self,
        start_date: datetime,
        end_date: datetime,
        data_fields: list[SleepDataFields] | None = None,
    ) -> list[SleepSeries]:
        """Get sleep."""
        data = {
            "action": "get",
            "startdate": start_date.timestamp(),
            "enddate": end_date.timestamp(),
        }
        if data_fields is not None:
            data["data_fields"] = ",".join(
                [str(sleep_data_field) for sleep_data_field in data_fields],
            )
        response = await self._request(
            "v2/sleep",
            data=data,
        )
        return [SleepSeries.from_api(sleep) for sleep in response["series"]]

    async def _get_sleep_summary(
        self,
        sleep_summary_data_fields: list[SleepSummaryDataFields] | None,
        base_data: dict[str, Any],
    ) -> list[SleepSummary]:
        data = {**base_data, "action": "getsummary"}
        if sleep_summary_data_fields is not None:
            data["data_fields"] = ",".join(
                [
                    str(sleep_data_field)
                    for sleep_data_field in sleep_summary_data_fields
                ],
            )
        response = await self._request(
            "v2/sleep",
            data=data,
        )
        return [SleepSummary.from_api(sleep) for sleep in response["series"]]

    async def get_sleep_summary_since(
        self,
        sleep_summary_since: datetime,
        sleep_summary_data_fields: list[SleepSummaryDataFields] | None = None,
    ) -> list[SleepSummary]:
        """Get all sleep summaries since sleep_summary_since."""
        return await self._get_sleep_summary(
            sleep_summary_data_fields,
            {"lastupdate": sleep_summary_since.timestamp()},
        )

    async def get_sleep_summary_in_period(
        self,
        start_date: date,
        end_date: date,
        sleep_summary_data_fields: list[SleepSummaryDataFields] | None = None,
    ) -> list[SleepSummary]:
        """Get sleep summary during period."""
        return await self._get_sleep_summary(
            sleep_summary_data_fields,
            {"startdateymd": str(start_date), "enddateymd": str(end_date)},
        )

    async def subscribe_notification(
        self,
        callback_url: str,
        notification_category: NotificationCategory,
    ) -> None:
        """Subscribe the callback_url for webhook updates."""
        await self._request(
            "notify",
            data={
                "action": "subscribe",
                "callbackurl": callback_url,
                "appli": notification_category,
            },
        )

    async def list_notification_configurations(
        self,
        notification_category: NotificationCategory | None = None,
    ) -> list[NotificationConfiguration]:
        """Subscribe the callback_url for webhook updates."""
        request_data: dict[str, Any] = {
            "action": "list",
        }
        if notification_category:
            request_data["appli"] = notification_category
        response = await self._request(
            "notify",
            data=request_data,
        )
        return [
            NotificationConfiguration.from_api(config)
            for config in response["profiles"]
        ]

    async def revoke_notification_configurations(
        self,
        callback_url: str,
        notification_category: NotificationCategory,
    ) -> None:
        """Revoke the configuration for webhook updates."""
        await self._request(
            "notify",
            data={
                "action": "revoke",
                "callbackurl": callback_url,
                "appli": notification_category,
            },
        )

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The WithingsClient object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
