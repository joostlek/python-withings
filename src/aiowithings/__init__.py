"""Asynchronous Python client for Withings."""
from .const import AUTHORIZATION_URL, TOKEN_URL
from .exceptions import (
    WithingsAuthenticationFailedError,
    WithingsBadStateError,
    WithingsConnectionError,
    WithingsError,
    WithingsErrorOccurredError,
    WithingsInvalidParamsError,
    WithingsTooManyRequestsError,
    WithingsUnauthorizedError,
    WithingsUnknownStatusError,
)
from .helpers import aggregate_measurements
from .models import (
    Activity,
    ActivityDataFields,
    ActivityDataOrigin,
    AuthScope,
    Device,
    DeviceBattery,
    DeviceModel,
    DeviceType,
    Goals,
    Measurement,
    MeasurementAttribution,
    MeasurementGroup,
    MeasurementGroupCategory,
    MeasurementType,
    NotificationCategory,
    Services,
    SleepDataFields,
    SleepSeries,
    SleepSeriesTimeData,
    SleepState,
    SleepSummary,
    SleepSummaryDataFields,
    WebhookCall,
    Workout,
    WorkoutCategory,
    WorkoutDataFields,
    get_measurement_type_from_notification_category,
)
from .withings import WithingsClient

__all__ = [
    "AUTHORIZATION_URL",
    "TOKEN_URL",
    "AuthScope",
    "ActivityDataFields",
    "ActivityDataOrigin",
    "Activity",
    "DeviceBattery",
    "Device",
    "DeviceType",
    "DeviceModel",
    "WithingsError",
    "WithingsConnectionError",
    "WithingsAuthenticationFailedError",
    "WithingsInvalidParamsError",
    "WithingsUnauthorizedError",
    "WithingsErrorOccurredError",
    "WithingsBadStateError",
    "WithingsTooManyRequestsError",
    "WithingsUnknownStatusError",
    "WithingsClient",
    "Goals",
    "MeasurementAttribution",
    "MeasurementGroupCategory",
    "MeasurementGroup",
    "MeasurementType",
    "Measurement",
    "get_measurement_type_from_notification_category",
    "aggregate_measurements",
    "NotificationCategory",
    "Services",
    "WebhookCall",
    "SleepSummary",
    "SleepSummaryDataFields",
    "SleepState",
    "SleepSeriesTimeData",
    "SleepSeries",
    "SleepDataFields",
    "WorkoutDataFields",
    "Workout",
    "WorkoutCategory",
]
