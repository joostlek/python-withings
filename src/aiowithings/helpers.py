"""Asynchronous Python client for Withings."""
from __future__ import annotations

from typing import TYPE_CHECKING

from aiowithings.models import MeasurementAttribution, MeasurementType, SleepSummary

if TYPE_CHECKING:
    from aiowithings import MeasurementGroup


def aggregate_measurements(
    measurements: list[MeasurementGroup],
) -> dict[MeasurementType, float]:
    """Aggregate the measurements to return a list of the latest measurements."""
    result: dict[MeasurementType, float] = {}

    groups = sorted(measurements, key=lambda group: group.taken_at)

    for measurement in groups:
        if measurement.attribution not in (
            MeasurementAttribution.UNKNOWN,
            MeasurementAttribution.DEVICE_ENTRY_FOR_USER_AMBIGUOUS,
        ):
            for data_point in measurement.measurements:
                result[data_point.measurement_type] = data_point.value

    return result


def aggregate_sleep_summary(
    sleep_summaries: list[SleepSummary],
) -> SleepSummary | None:
    """Aggregate the sleep summaries to return a combined report."""
    if len(sleep_summaries) == 0:
        return None
    if len(sleep_summaries) == 1:
        return sleep_summaries[0]
    sorted_sleep_summaries = sorted(
        sleep_summaries,
        key=lambda sleep_summary: sleep_summary.start_date,
    )

    def average(data: list[int]) -> int:
        return round(sum(data) / len(data))

    return SleepSummary(
        start_date=sorted_sleep_summaries[0].start_date,
        end_date=sorted_sleep_summaries[1].end_date,
        date=sorted_sleep_summaries[0].date,
        hashed_device_id=sorted_sleep_summaries[0].hashed_device_id,
        breathing_disturbances_intensity=average(
            [
                summary.breathing_disturbances_intensity
                for summary in sleep_summaries
                if summary.breathing_disturbances_intensity is not None
            ],
        ),
        deep_sleep_duration=sum(
            summary.deep_sleep_duration
            for summary in sleep_summaries
            if summary.deep_sleep_duration is not None
        ),
        sleep_latency=average(
            [
                summary.sleep_latency
                for summary in sleep_summaries
                if summary.sleep_latency is not None
            ],
        ),
        wake_up_latency=average(
            [
                summary.wake_up_latency
                for summary in sleep_summaries
                if summary.wake_up_latency is not None
            ],
        ),
        average_heart_rate=average(
            [
                summary.average_heart_rate
                for summary in sleep_summaries
                if summary.average_heart_rate is not None
            ],
        ),
        max_heart_rate=average(
            [
                summary.max_heart_rate
                for summary in sleep_summaries
                if summary.max_heart_rate is not None
            ],
        ),
        min_heart_rate=average(
            [
                summary.min_heart_rate
                for summary in sleep_summaries
                if summary.min_heart_rate is not None
            ],
        ),
        light_sleep_duration=sum(
            summary.light_sleep_duration
            for summary in sleep_summaries
            if summary.light_sleep_duration is not None
        ),
        rem_sleep_duration=sum(
            summary.rem_sleep_duration
            for summary in sleep_summaries
            if summary.rem_sleep_duration is not None
        ),
        average_respiration_rate=average(
            [
                summary.average_respiration_rate
                for summary in sleep_summaries
                if summary.average_respiration_rate is not None
            ],
        ),
        max_respiration_rate=average(
            [
                summary.max_respiration_rate
                for summary in sleep_summaries
                if summary.max_respiration_rate is not None
            ],
        ),
        min_respiration_rate=average(
            [
                summary.min_respiration_rate
                for summary in sleep_summaries
                if summary.min_respiration_rate is not None
            ],
        ),
        sleep_score=max(
            (
                summary.sleep_score
                for summary in sleep_summaries
                if summary.sleep_score is not None
            ),
            default=0,
        ),
        snoring=average(
            [
                summary.snoring
                for summary in sleep_summaries
                if summary.snoring is not None
            ],
        ),
        snoring_count=sum(
            summary.snoring_count
            for summary in sleep_summaries
            if summary.snoring_count is not None
        ),
        wake_up_count=sum(
            summary.wake_up_count
            for summary in sleep_summaries
            if summary.wake_up_count is not None
        ),
        total_time_awake=average(
            [
                summary.total_time_awake
                for summary in sleep_summaries
                if summary.total_time_awake is not None
            ],
        ),
        active_movement_duration=None,
        apnea_hypopnea_index=None,
        average_movement_score=None,
        external_time_asleep=None,
        out_of_bed_count=None,
        rem_sleep_phase_count=None,
        sleep_efficiency=None,
        time_awake_during_sleep=None,
        total_sleep_time=None,
        total_time_in_bed=None,
        withings_index=None,
    )
