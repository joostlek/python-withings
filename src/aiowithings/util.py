"""Asynchronous Python client for Withings."""
from __future__ import annotations

from typing import Any, TypeVar, cast

from aiowithings.const import LOGGER

_EnumT = TypeVar("_EnumT")


def to_enum(
    enum_class: type[_EnumT],
    value: Any,
    default_value: _EnumT,
) -> _EnumT:
    """Convert a value to an enum and log if it doesn't exist."""
    try:
        return enum_class(value)  # type: ignore[call-arg]
    except ValueError:
        LOGGER.warning(
            "%s is an unsupported value for %s, please report this at https://github.com/joostlek/python-withings/issues",
            value,
            str(enum_class),
        )
        return default_value


def get_measurement(value: int, unit: int) -> float:
    """Convert value and unit to real value."""
    return cast(float, value * pow(10, unit))


def get_measurement_from_dict(input_dict: dict[str, Any]) -> float:
    """Convert dict with value and unit to real value."""
    value = cast(int, input_dict["value"])
    unit = cast(int, input_dict["unit"])
    return get_measurement(value, unit)
