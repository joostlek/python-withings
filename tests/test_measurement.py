"""Asynchronous Python client for Withings."""
from datetime import datetime, timezone

import aiohttp
from aresponses import ResponsesMockServer

from aiowithings import WithingsClient
from aiowithings.models import MeasurementType
from syrupy import SnapshotAssertion

from . import load_fixture

WITHINGS_URL = "wbsapi.withings.net"


async def test_get_measurement_since(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving measurements."""
    aresponses.add(
        WITHINGS_URL,
        "/measure",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("measurement.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        withings = WithingsClient(session=session)
        withings.authenticate("test")
        response = await withings.get_measurement_since(
            datetime.now(tz=timezone.utc),
            measurement_types=[MeasurementType.WEIGHT],
        )
        assert response == snapshot
        await withings.close()


async def test_get_measurement_period(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving measurements."""
    aresponses.add(
        WITHINGS_URL,
        "/measure",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("measurement.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        withings = WithingsClient(session=session)
        withings.authenticate("test")
        response = await withings.get_measurement_in_period(
            start_date=datetime.now(tz=timezone.utc),
            end_date=datetime.now(tz=timezone.utc),
        )
        assert response == snapshot
        await withings.close()
