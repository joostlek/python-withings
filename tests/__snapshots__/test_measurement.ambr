# serializer version: 1
# name: test_get_measurement_period
  list([
    dict({
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'category': <MeasurementGroupCategory.REAL: 1>,
      'device_id': 'f998be4b9ccc9e136fd8cd8e8e344c31ec3b271d',
      'group_id': 4815757309,
      'hashed_device_id': 'f998be4b9ccc9e136fd8cd8e8e344c31ec3b271d',
      'measurements': list([
        dict({
          'measurement_type': <MeasurementType.WEIGHT: 1>,
          'value': 118.003,
        }),
      ]),
      'stored_at': datetime.datetime(2023, 9, 2, 10, 39, 45, tzinfo=datetime.timezone.utc),
      'taken_at': datetime.datetime(2023, 9, 2, 10, 39, 11, tzinfo=datetime.timezone.utc),
      'updated_at': datetime.datetime(2023, 9, 2, 10, 39, 45, tzinfo=datetime.timezone.utc),
    }),
    dict({
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'category': <MeasurementGroupCategory.REAL: 1>,
      'device_id': 'f998be4b9ccc9e136fd8cd8e8e344c31ec3b271d',
      'group_id': 4815757309,
      'hashed_device_id': 'f998be4b9ccc9e136fd8cd8e8e344c31ec3b271d',
      'measurements': list([
        dict({
          'measurement_type': <MeasurementType.WEIGHT: 1>,
          'value': 118.003,
        }),
      ]),
      'stored_at': datetime.datetime(2023, 9, 2, 10, 39, 45, tzinfo=datetime.timezone.utc),
      'taken_at': datetime.datetime(2023, 9, 2, 10, 39, 11, tzinfo=datetime.timezone.utc),
      'updated_at': datetime.datetime(2023, 9, 2, 10, 39, 45, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_get_measurement_since
  list([
    dict({
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'category': <MeasurementGroupCategory.REAL: 1>,
      'device_id': 'f998be4b9ccc9e136fd8cd8e8e344c31ec3b271d',
      'group_id': 4815757309,
      'hashed_device_id': 'f998be4b9ccc9e136fd8cd8e8e344c31ec3b271d',
      'measurements': list([
        dict({
          'measurement_type': <MeasurementType.WEIGHT: 1>,
          'value': 118.003,
        }),
      ]),
      'stored_at': datetime.datetime(2023, 9, 2, 10, 39, 45, tzinfo=datetime.timezone.utc),
      'taken_at': datetime.datetime(2023, 9, 2, 10, 39, 11, tzinfo=datetime.timezone.utc),
      'updated_at': datetime.datetime(2023, 9, 2, 10, 39, 45, tzinfo=datetime.timezone.utc),
    }),
    dict({
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'category': <MeasurementGroupCategory.REAL: 1>,
      'device_id': 'f998be4b9ccc9e136fd8cd8e8e344c31ec3b271d',
      'group_id': 4815757309,
      'hashed_device_id': 'f998be4b9ccc9e136fd8cd8e8e344c31ec3b271d',
      'measurements': list([
        dict({
          'measurement_type': <MeasurementType.WEIGHT: 1>,
          'value': 118.003,
        }),
      ]),
      'stored_at': datetime.datetime(2023, 9, 2, 10, 39, 45, tzinfo=datetime.timezone.utc),
      'taken_at': datetime.datetime(2023, 9, 2, 10, 39, 11, tzinfo=datetime.timezone.utc),
      'updated_at': datetime.datetime(2023, 9, 2, 10, 39, 45, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
